[CGP-374] Added dynamic built-in types

effectfully opened the pull request:
This is preliminary, don't look here yet.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs:
```diff
@@ -79,6 +82,10 @@ instance ( PrettyBy config (Constant ())
         , prettyBy config arg
         ]
+    prettyBy config (UnreadableBuiltinConstAppError arg)    = fold
+        [ "A built-in couldn't be converted from Plutus Core to Haskell:", "\n"
```
I'd prefer to say something like "Could not construct denotation for a built-in" or something that's less implementation-language specific.

effectfully replied:
I've changed that, but I think the error message is now slightly more cryptic.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/DynamicType.hs:
```diff
@@ -0,0 +1,89 @@
+recursive instances as shown above.
+
+Instance resolution is driven by Haskell types rather than names of PLC types.
+An alternative design would be something like
```
I'd appreciate a bit more discussion of the *actual* design before contrasting it with alternatives!

effectfully replied:
I've added a few lines, but I'm not sure what to describe there. Here are the lines:

```
An @KnownDynamicBuiltinType dyn@ instance provides

1. a way to encode @dyn@ as a PLC type ('getTypeEncoding')
2. a function that encodes values of type @dyn@ as PLC terms ('makeDynamicBuiltin')
3. a function that decodes PLC terms back to Haskell values ('readDynamicBuiltin')

The last two are ought to constitute an isomorphism (modulo 'Quote' and 'Maybe').
```

Is there anything else that needs to be described? I mean, this stuff is not complicated.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/DynamicType.hs:
```diff
@@ -0,0 +1,89 @@
+-- | Haskell types known to exist on the PLC side.
+class KnownDynamicBuiltinType dyn where
+    -- | The name of @dyn@ used on the PLC side.
+    dynamicBuiltinType :: proxy dyn -> DynamicBuiltinType
```
Is there a reason not to use `Data.Proxy` here? Clearly you can in fact use anything you want for `proxy`, but I've not seen it done this way before and I was wondering if there's a reason not to.

michaelpj replied:
Ah I see, cute, this means you can use the `DynamicBuiltinType` as a proxy for `dyn`. Very nice.

michaelpj replied:
And indeed, since `dyn` is going to be existentially quantified, you might struggle to declare a `Proxy` of the right type.

effectfully replied:
This is a fairly standard trick that people from the hasochistic (the terms originates [here](http://homepages.inf.ed.ac.uk/slindley/papers/hasochism.pdf)) camp use. I do not have a reference, though.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/DynamicType.hs:
```diff
@@ -0,0 +1,89 @@
+
+    -- | Convert a Haskell value to the corresponding PLC value.
+    -- 'Nothing' represents a conversion failure.
+    makeDynamicBuiltin :: dyn -> Maybe (Value TyName Name ())
```
Should this produce something in `Quote`?

effectfully replied:
I asked this question to myself. Not sure. I'll add `Quote`.

effectfully replied:
Done.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs:
```diff
@@ -148,6 +149,7 @@ data BuiltinType size
 data TypedBuiltin size a where
     TypedBuiltinSized :: SizeEntry size -> TypedBuiltinSized a -> TypedBuiltin size a
     TypedBuiltinBool  :: TypedBuiltin size Bool
+    TypedBuiltinDyn   :: KnownDynamicBuiltinType dyn => TypedBuiltin size dyn
```
There's those existentials!

effectfully replied:
Yes, but they are not in the AST itself, so we still have `Generic`, `NFData`, etc instances for `Term`. We're not even complicating what we already had, because the typed built-ins machinery was already complicated enough (it uses GADTs after all which are strictly more powerful than existentials).

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs:
```diff
@@ -243,6 +243,7 @@ instance Eq size => GEq (TypedBuiltin size) where
         guard $ size1 == size2
         tbs1 `geq` tbs2
     TypedBuiltinBool             `geq` TypedBuiltinBool             = Just Refl
+    -- A typed built-in is not equal to itself. TODO: do something about it.
```
We could require an `Eq` instance as a superclass constraint on `KnownDynamicBuiltin`?

effectfully replied:
Throwing errors now.

michaelpj commented on language-plutus-core/src/Language/PlutusCore/Lexer/Type.hs:
```diff
@@ -89,11 +91,19 @@ which has the desired type signature:
--- | The type of dynamic builtin functions. I.e. functions that exist on certain chains and do
+-- | The type of dynamic built-in types. I.e. types that exist on certain chains and do
+-- not exist on others. Each 'DynamicBuiltinType' has an associated kind --
+-- this allows to kind check dynamic built-in types just like static ones.
```
This is confusing - reading this comment I expected there to be a kind here, but there isn't.

michaelpj replied:
And, indeed, given your note this seems to be false and they all have kind `*` (and that's what the typechecker says).

effectfully replied:
Oh yes, sorry, old comment.

michaelpj commented on plutus-core-interpreter/test/DynamicBuiltins/Char.hs:
```diff
@@ -0,0 +1,72 @@
+-- the contents of an external 'IORef'. In the end read the 'IORef', reverse its contents and check
+-- that you got the exact same sequence of 'Char's that was originally generated.
+-- Calls 'unsafePerformIO' internally while evaluating the term, because the CEK machine can only handle
+-- pure things and 'unsafePerformIO' is the way to pretend an effecful thing is pure.
```
Impressively vile, but effective.

effectfully replied:
True :)

vmchale commented on .hlint.yaml:
```diff
@@ -1,7 +1,7 @@
-  - {name: unsafePerformIO, within: [PlutusPrelude, Language.PlutusCore.Generators.Internal.Entity, Language.Plutus.CoreToPLC.Plugin]}
-  - {name: error, within: [Main, PlutusPrelude, Language.PlutusCore.StdLib.Meta, Evaluation.Constant.Success, Language.PlutusCore.Constant.Apply, Language.PlutusCore.Constant.Typed, Language.PlutusCore.Evaluation.CkMachine, Language.PlutusCore.TypeSynthesis, Language.PlutusCore.Generators.Internal.Entity, Language.PlutusCore.Generators.Internal.Utils, Language.PlutusCore.Constant.Make, Language.P ...
+  - {name: unsafePerformIO, within: [PlutusPrelude, Language.PlutusCore.Generators.Internal.Entity, DynamicBuiltins.OnEach, Language.Plutus.CoreToPLC.Plugin]}
+  - {name: error, within: [Main, PlutusPrelude, Language.PlutusCore.StdLib.Meta, Evaluation.Constant.Success, Language.PlutusCore.Constant.Apply, Language.PlutusCore.Constant.Typed, Language.PlutusCore.Evaluation.CkMachine, Language.PlutusCore.TypeSynthesis, Language.PlutusCore.Generators.Internal.Entity, Language.PlutusCore.Generators.Internal.Utils, Language.PlutusCore.Constant.Make, DynamicBui ...
```
We can probably stop banning `error` altogether, no?

effectfully replied:
If anyone agrees, I can remove it. @michaelpj, what do you think?

michaelpj replied:
Hm, `error` is still usually a code smell, so I somewhat like having the inconvenience of having to put it in the hlint file.

effectfully replied:
I won't touch it then.

vmchale commented on language-plutus-core/src/Language/PlutusCore/Constant/Apply.hs:
```diff
@@ -79,6 +82,10 @@ instance ( PrettyBy config (Constant ())
         , prettyBy config arg
         ]
+    prettyBy config (UnreadableBuiltinConstAppError arg)    = fold
+        [ "Could not construct denotation for a built-in:", "\n"
```
If you're using a pretty-printer, you probably want to use something like `hardline` instead of `"\n"`.

effectfully replied:
Yeah. But there are other pieces with `"\n"`, I don't want to refactor stuff in this PR. I'll do that later, thanks.

effectfully commented:
This PR only adds the basic functionality. It'll be followed by another PR that does quite more complicated things (but doesn't change the interface. Well, almost). In the final PR we'll decide on the interface and how this entire thing should be used. The full work is not done yet (though, I've just got strings it seems, but that's for the another PR)

I redesigned the machinery a bit: removed `DynamicBuiltinType` and made dynamic built-in types type aliases rather than `newtype`s. The reason for that is that I started type checking tests and realized that I very stupidly implemented names for type-level newtype wrappers, but nothing for term-level wrappers and unwrappers. I do not think having newtypes is important for us at this point, so I just implemented the simplest thing that works.

I'll add a bit more docs, squash-merge and move to preparing the subsequent PR if that's fine.

mchakravarty commented:
> I'll add a bit more docs, squash-merge and move to preparing the subsequent PR if that's fine.

👍🏼
